File: solutions/lua/flower-field/1/flower-field.lua
local function annotate(garden)
	local res = {}
	local height = #garden
	local width = garden[1] and #garden[1] or 0
	local function at(i, j)
		if i < 1 or i > height or j < 1 or j > width or garden[i]:sub(j, j) == ' ' then
			return 0
		else
			return 1
		end
	end
	for i = 1, height do
		local row = ""
		for j = 1, width do
			local neighbours =
				at(i-1, j-1) + at(i-1, j) + at(i-1, j+1) +
				at(i, j-1) + at(i, j+1) +
				at(i+1, j-1) + at(i+1, j) + at(i+1, j+1)
			if garden[i]:sub(j, j) == '*' then
				row = row .. '*'
			elseif neighbours == 0 then
				row = row .. ' '
			else
				row = row .. tostring(neighbours)
			end
		end
		res[i] = row
	end
	return res
end

return { annotate = annotate }

File: solutions/c/flower-field/1/flower_field.c
#include "flower_field.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef struct { int first; int second; } pair;

static pair offsets[8] = {
	{-1, -1}, {-1, 0}, {-1, 1},
	{0, -1},           {0, 1},
	{1, -1},  {1, 0},  {1, 1},
};

char **annotate(const char **garden, const size_t rows) {
	if (rows == 0) return 0x0;
	char **res = malloc(sizeof(char*) * 100);
	int sz2 = 0;
	int height = rows;
	int width = (rows > 0) ? strlen(garden[0]) : 0;
	for (int i = 0; i < height; i++) {
		char *row = malloc(sizeof(char) * 100);
		int sz = 0;
		for (int j = 0; j < width; j++) {
			int neighbours = 0;
			for (int k = 0; k < 8; k++) {
				pair p = offsets[k];
				int dx = p.first;
				int dy = p.second;
				int ni = i+dx, nj = j+dy;
				neighbours += ! (ni < 0 || ni > (height-1) || nj < 0 || nj > (width-1) || garden[ni][nj] == ' ');
			}
			if (garden[i][j] == '*')
				row[sz++] = '*', row[sz] = '\0';
			else if (neighbours == 0)
				row[sz++] = ' ', row[sz] = '\0';
			else {
				char buf[50];
				sprintf(buf, "%d", neighbours);
				strcat(row, buf);
				sz = strlen(row);
			}
		}
		res[sz2++] = row;
		res[sz2] = 0x0;
	}
	return res;
}

void free_annotation(char **annotation) {
	(void)annotation;
}
